### PR TITLE
Add encryption support to OpenSSL

### PR DESCRIPTION
**Pull request by bkuschel**
Add missing encryption feature when OpenSSL is used with ClickHouse.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
N/A

### CODE REVIEW

File: src/Compression/CompressionCodecEncrypted.cpp [hunk 1]
```diff
@@ -186,6 +191,138 @@ size_t decrypt(std::string_view ciphertext, char * plaintext, EncryptionMethod m
...
+    const int ok_tag = EVP_CIPHER_CTX_ctrl(encrypt_ctx,
+                                           EVP_CTRL_GCM_GET_TAG,
+                                           tag_size,
+                                           reinterpret_cast<uint8_t *>(ciphertext_and_tag) + plaintext.size());
+    if (!ok_tag)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    EVP_CIPHER_CTX_free(encrypt_ctx);
+
+    return plaintext.size() + tag_size;
```
Comment (yakov-olkhovskiy) [inline src/Compression/CompressionCodecEncrypted.cpp:267]: Are you sure the encrypted size is equal to input text size? shouldn't it be padded to block's multiple plus some bytes for IV? maybe me should use returned size in `out_len`?

File: src/Compression/CompressionCodecEncrypted.cpp [hunk 2]
```diff
@@ -186,6 +191,138 @@ size_t decrypt(std::string_view ciphertext, char * plaintext, EncryptionMethod m
...
+
+    const int ok_decryptfinal = EVP_DecryptFinal_ex(decrypt_ctx,
+                                                    reinterpret_cast<uint8_t *>(plaintext) + out_len,
+                                                    reinterpret_cast<int32_t *>(&out_len));
+    if (!ok_decryptfinal)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    EVP_CIPHER_CTX_free(decrypt_ctx);
+
+    return ciphertext.size() - tag_size;
```
Comment (yakov-olkhovskiy) [inline src/Compression/CompressionCodecEncrypted.cpp:323]: the same as above - why we don't use `out_len`? - isn't it more reliable?

File: src/Compression/CompressionCodecEncrypted.cpp (hunk 1 above)
Comment (bkuschel) [reply src/Compression/CompressionCodecEncrypted.cpp:267]: I wanted to do this initially but in an effort to try to make the encryption as compatible as possible to boringssl, I used the same method for dealing with lengths as boringSSL:
https://github.com/google/boringssl/blob/master/crypto/fipsmodule/cipher/aead.c#L141

File: src/Compression/CompressionCodecEncrypted.cpp (hunk 2 above)
Comment (bkuschel) [reply src/Compression/CompressionCodecEncrypted.cpp:323]: same as above:
https://github.com/google/boringssl/blob/master/crypto/fipsmodule/cipher/aead.c#L219

File: src/Compression/CompressionCodecEncrypted.cpp (hunk 1 above)
Comment (yakov-olkhovskiy) [reply src/Compression/CompressionCodecEncrypted.cpp:267]: I don't think it's a good idea - boringssl's writers know internals they are using so they are free to do whatever they consider suitable, however as a library users we are not free to make any such assumptions - there is a paragraphs on openssl doc dedicated to peculiarity of output length:
https://www.openssl.org/docs/man3.0/man3/EVP_EncryptUpdate.html

Comment (bkuschel) [reply src/Compression/CompressionCodecEncrypted.cpp:267]: ok, let me see if I can make it work with out length. There is the intricacies of the tag being appended onto the length. I just want to make sure that the SSL providers can decrypt each other encryption.

File: src/Compression/CompressionCodecEncrypted.cpp [hunk 3]
```diff
@@ -186,6 +191,138 @@ size_t decrypt(std::string_view ciphertext, char * plaintext, EncryptionMethod m
...
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    const int ok_nonceinit = EVP_EncryptInit_ex(encrypt_ctx, nullptr, nullptr,
+                                                reinterpret_cast<const uint8_t*>(key.data()),
+                                                reinterpret_cast<const uint8_t *>(nonce.data()));
+    if (!ok_nonceinit)
+        throw Exception(lastErrorString(), ErrorCodes::OPENSSL_ERROR);
+
+    size_t out_len;
+    const int ok_encryptudpate = EVP_EncryptUpdate(encrypt_ctx,
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecEncrypted.cpp:243]: `ok_encryptupdate`?

**Review by rschu1ze — COMMENTED** [review]
I generally like that the special handling for open/boringssl is moved from all-over-the-place to this file.

Caveat is that I am not an expert in boring/openSSL and can't verify if the encryption/decryption invocation was done right (it looks okay).

**Comment by bkuschel** [comment]
@yakov-olkhovskiy @rschu1ze I rebased the new exception changes. I changed the size to the one returned by the function (ends up being the same) and fixed the potential leak by using a catch block.  Is good now?

File: src/Compression/CompressionCodecEncrypted.cpp [hunk 4]
```diff
@@ -186,6 +191,138 @@ size_t decrypt(std::string_view ciphertext, char * plaintext, EncryptionMethod m
...
+    }
+}
+
+/// Encrypt plaintext with particular algorithm and put result into ciphertext_and_tag.
+/// This function get key and nonce and encrypt text with their help.
+/// If something went wrong (can't init context or can't encrypt data) it throws exception.
+/// It returns length of encrypted text.
+size_t encrypt(std::string_view plaintext, char * ciphertext_and_tag, EncryptionMethod method, const String & key, const String & nonce)
+{
+    EVP_CIPHER_CTX *encrypt_ctx;
```
[the remark this answers is not in the document]
Comment (bkuschel) [reply src/Compression/CompressionCodecEncrypted.cpp:220]: I'll free it in a catch block.

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
ClickHouse Stateless Tests (debug) [2/5]:
- 00738_lock_for_inner_table: infra issue

FunctionalStatefulTestUBsanParallelReplicas:
- infra issue

ClickHouse Integration Tests (tsan) [6/6]:
- test_backup_restore_new: opened issue #46056

**Pull request merged**